crates/oxc_linter/src/plugin.rs

```diff
@@ -0,0 +1,213 @@
+use std::{env, fs, path::PathBuf, rc::Rc, sync::Arc};
+
+use oxc_diagnostics::miette::{miette, LabeledSpan};
+use oxc_semantic::Semantic;
+use trustfall::{execute_query, Schema, TryIntoStruct};
+
+use crate::{
+    context::LintContext,
+    lint_adapter::{InputQuery, LintAdapter},
+};
+
+enum SpanInfo {
+    SingleSpanInfo(SingleSpanInfo),
+    MultipleSpanInfo(MultipleSpanInfo),
+}
+
+#[derive(Debug, serde::Deserialize)]
+struct SingleSpanInfo {
+    span_start: u64,
+    span_end: u64,
+}
+
+#[derive(Debug, serde::Deserialize)]
+struct MultipleSpanInfo {
+    span_start: Box<[u64]>,
+    span_end: Box<[u64]>,
+}
+
+pub struct LinterPlugin {
+    pub(super) rules: Vec<InputQuery>,
+    schema: Schema,
+}
+
+impl LinterPlugin {
+    pub fn new() -> Self {
+        let queries_path = env::var("OXC_PLUGIN").unwrap();
+        let rules = fs::read_dir(queries_path)
+            .expect("to readdir queries_path folder")
```

Comment:
This whole thing is really brittle, I would prefer if we just did a recursive search for rules from a certain rules folder.